### ISSUE TITLE
Magit is broken when working in a remote directory and the local git is a cygwin git

### ISSUE THREAD

**Issue by montag451** [magit/magit#4976]
The problem seems to come from `magit-expand-git-file-name`. When the local git is a cygwin git, `magit-cygwin-mount-points` is not nil and then the path is prefixed with a Windows path (a local one) while the `default-directory` is a remote one. For example if the local git is coming from a MSYS2 installation located in `C:\msys64` and the remote path of the Git repository is `/home/user/foo` (without the TRAMP part) then `magit-expand-git-file-name` will return `C:/msys64/home/user/foo` which is wrong.

**Comment by tarsius** [comment]
I don't understand this code and I don't have access to Windows to try it out.

Could you post the value of `magit-cygwin-mount-points`. It's possible that would give me some useful insight.

**Comment by montag451** [comment]
On my machine the value of `magit-cygwin-mount-points` is:
``` elisp
(("/bin/" . "c:/msys64/usr/bin/")
 ("/c/" . "c:/")
 ("/l/" . "l:/")
 ("/o/" . "o:/")
 ("/r/" . "r:/")
 ("/y/" . "y:/")
 ("/" . "c:/msys64/"))
```

**Comment by montag451** [comment]
I think the fix should be to just skip the dance with `magit-cygwin-mount-points` when the `default-directory` is a remote one. Something like that:
```elisp
(defun magit-expand-git-file-name (filename)
  (unless (file-name-absolute-p filename)
    (setq filename (expand-file-name filename)))
  (unless (file-remote-p default-directory)
    (if-let ((cyg:win (cl-assoc filename magit-cygwin-mount-points
                                :test (lambda (f cyg) (string-prefix-p cyg f)))))
        (concat (cdr cyg:win)
                (substring filename (length (car cyg:win))))
      filename)))
```

**Comment by tarsius** [comment]
That does sound reasonable, but of course the remote machine may also use cygwin. Then again, the local mounts would likely differ from those on the remote...

I'll install that, and if someone reports that it breaks their workflow, we can worry about it then. Because I never used Windows for anything but running games and a browser, I don't really know what's going on. (The other problem being that there are many, configurable ways of running unixy things on Windows.)

**Comment by montag451** [comment]
It sounds like a reasonable solution. I think that the likelihood of someone connecting to a remote machine which uses a cygwin git is far less that someone using a local cygwin git and connecting to a remote machine which uses a non-cygwin git.

**Comment by mickaelfiorentino** [comment]
Hello, I just want to report that this issue does break my workflow. 

For reasons I will not bother you with, instead of using magit within my Linux machine on my git repo, I had to use magit from my windows machine on the same git repo mounted as a network drive (cifs, samba) ; something like: "//server/path/to/repo" 

magit-git-executable points to my MSYS2 install : "c:/msys2/usr/lib/git-core/git.exe"
from within Emacs Dired inside my repo "//server/path/to/repo", the command
(magit-expand-git-file-name (magit-rev-parse-safe "--show-toplevel"))
returns "c:/msys2//server/path/to/repo" 
The result is that magit-status does not work since this path does not exist.

The code of magit-expand-git-file-name (from where I found this issue) is a bit too complex for me to propose a viable alternative implementation.

Thank you for your work, Magit is awesome.
